### PR TITLE
Error Handling

### PR DESCRIPTION
**Pull request by rvashurin**
Goal of this PR is to keep information about exception even when they are ignored, and present info about failures in a convenient way after the calculation is done.

### CODE REVIEW

File: src/lm_polygraph/utils/dataset.py [hunk 1]
```diff
@@ -165,16 +165,12 @@ def from_datasets(
...
-                if x_column == "fr"
-                else "English"
+                else "French" if x_column == "fr" else "English"
             )
             target_lang = (
                 "German"
                 if y_column == "de"
-                else "French"
-                if y_column == "fr"
-                else "English"
```
Comment (rvashurin) [inline src/lm_polygraph/utils/dataset.py:177]: Linter (black) fixes, not important

File: src/lm_polygraph/utils/manager.py [hunk 2]
```diff
@@ -328,6 +329,7 @@ def __init__(
         self.gen_metrics: Dict[Tuple[str, str], List[float]] = defaultdict(list)
         self.estimations: Dict[Tuple[str, str], List[float]] = defaultdict(list)
         self.metrics: Dict[Tuple[str, str, str, str], float] = {}
+        self.total_bad_estimators: Dict[Estimator, float] = {}
```
Comment (rvashurin) [inline src/lm_polygraph/utils/manager.py:332]: Keep track of all estimators that failed.

File: src/lm_polygraph/utils/manager.py [hunk 3]
```diff
@@ -457,6 +460,7 @@ def __call__(self) -> Dict[Tuple[str, str, str, str], float]:
                     key = (bad_estimator.level, str(bad_estimator))
                     self.estimations.pop(key, None)
                     self.ensemble_estimators.remove(bad_estimator)
+                    self.total_bad_estimators[bad_estimator] = batch_i
```
Comment (rvashurin) [inline src/lm_polygraph/utils/manager.py:463]: save them along with batch number on which they failed

File: src/lm_polygraph/utils/manager.py [hunk 4]
```diff
@@ -488,15 +492,15 @@ def __call__(self) -> Dict[Tuple[str, str, str, str], float]:
                         oracle_score = ue_metric(-rec_metric, rec_metric)
                         random_score = get_random_scores(ue_metric, rec_metric)
                         ue_metric_val = ue_metric(rec_ue, rec_metric)
-                        self.metrics[
-                            e_level, e_name, gen_name, str(ue_metric)
-                        ] = ue_metric_val
+                        self.metrics[e_level, e_name, gen_name, str(ue_metric)] = (
+                            ue_metric_val
+                        )
```
Comment (rvashurin) [inline src/lm_polygraph/utils/manager.py:497]: Linter changes (black), not important

File: src/lm_polygraph/utils/manager.py [hunk 5]
```diff
@@ -520,9 +524,11 @@ def calculate(self, batch_stats: dict, calculators: list, inp_texts: list) -> di
                     batch_stats[stat] = stat_value
             except Exception as e:
                 if self.ignore_exceptions:
-                    log_msg = f"Caught exception while calculating stats: {e} in Stat Calculator {stat_calculator}"
+                    lineno = e.__traceback__.tb_lineno
+                    log_msg = f"Caught exception while calculating stats: {e} in Stat Calculator {stat_calculator}, line {lineno}. Expect dependent estimator to fail.\n"
+                    sys.stderr.write("\n\n")
                     sys.stderr.write(log_msg)
-                    print(log_msg)
+                    sys.stderr.write(traceback.format_exc())
```
Comment (rvashurin) [inline src/lm_polygraph/utils/manager.py:531]: If error occurs and we are ignoring exceptions, dump full error traceback and message about exception into stderr.

File: src/lm_polygraph/utils/manager.py [hunk 6]
```diff
@@ -550,9 +556,11 @@ def estimate(
             except Exception as e:
                 if self.ignore_exceptions:
                     bad_estimators.append(estimator)
-                    log_msg = f"Caught exception while estimating uncertainty: {e} in estimator {estimator}. Estimator will be removed."
+                    lineno = e.__traceback__.tb_lineno
+                    log_msg = f"Caught exception while estimating uncertainty: {e} in estimator {estimator}, line {lineno}. Estimator will be removed.\n"
+                    sys.stderr.write("\n\n")
                     sys.stderr.write(log_msg)
-                    print(log_msg)
+                    sys.stderr.write(traceback.format_exc())
```
Comment (rvashurin) [inline src/lm_polygraph/utils/manager.py:563]: Same for estimators

File: src/lm_polygraph/utils/processor.py [hunk 7]
```diff
@@ -76,13 +77,21 @@ def on_batch(
...
         print("=" * 50 + " METRICS " + "=" * 50)
         print("Metrics:")
         print()
         for key, val in metrics.items():
             print(f"{key}: {val}")
             print()
+        if len(bad_estimators) > 0:
+            print("=" * 45 + " FAILED ESTIMATORS " + "=" * 45)
+            for bad_estimator, batch_i in bad_estimators.items():
+                print(str(bad_estimator), " on batch ", batch_i)
```
Comment (rvashurin) [inline src/lm_polygraph/utils/processor.py:97]: On finished eval dump info about all failed estimators and batch id on which they failed.

File: src/lm_polygraph/utils/token_restoration.py [hunk 8]
```diff
@@ -256,9 +256,9 @@ def collect_token_level_uncertainties(
                             ue = output_uncertainties_reshaped[key][
                                 obs_id, beam_id, _iter
                             ]
-                        token_level_uncertainties[key][
-                            obs_id, seq_i, _iter
-                        ] = torch.tensor(ue)
+                        token_level_uncertainties[key][obs_id, seq_i, _iter] = (
+                            torch.tensor(ue)
+                        )
```
Comment (rvashurin) [inline src/lm_polygraph/utils/token_restoration.py:261]: Linter changes

**Review by SergeyPetrakov — APPROVED** [review]

**Pull request merged**